src/libraries/System.Text.Json/source_generation/src/JsonSourceGeneratorHelper.Generate.cs

```diff
@@ -0,0 +1,676 @@
+                    {
+                        _executionContext.AddSource(
+                            metadataFileName,
+                            SourceText.From(GenerateForCollection(typeMetadata), Encoding.UTF8));
+
+                        GenerateTypeMetadata(typeMetadata.CollectionKeyTypeMetadata);
+                        GenerateTypeMetadata(typeMetadata.CollectionValueTypeMetadata);
+                    }
+                    break;
+                case ClassType.Object:
+                    {
+                        // TODO: this codepath assumes deserialization with a public parameterless ctor.
+                        // Add mechanism to detect otherwise and opt-in for runtime metadata generation and/or serialization
+                        // with JsonSerializer's dynamic code paths.
+
+                        _executionContext.AddSource(
+                            metadataFileName,
+                            SourceText.From(GenerateForObject(typeMetadata), Encoding.UTF8));
+
+                        if (typeMetadata.PropertiesMetadata != null)
+                        {
+                            foreach (PropertyMetadata metadata in typeMetadata.PropertiesMetadata)
+                            {
+                                GenerateTypeMetadata(metadata.TypeMetadata);
+                            }
+                        }
+                    }
+                    break;
+                case ClassType.TypeUnsupportedBySourceGen:
+                    {
+                        _executionContext.ReportDiagnostic(
+                            Diagnostic.Create(TypeNotSupported, Location.None, new string[] { typeMetadata.CompilableName }));
+                        return;
+                    }
+                default:
+                    {
+                        throw new InvalidOperationException();
+                    }
+            }
+        }
+
+        private string GenerateForTypeWithKnownConverter(TypeMetadata typeMetadata)
+        {
+            string typeCompilableName = typeMetadata.CompilableName;
+            string typeFriendlyName = typeMetadata.FriendlyName;
+
+            string metadataInitSource = $@"_{typeFriendlyName} = {JsonMetadataServicesClassName}.{GetCreateValueInfoMethodRef(typeCompilableName)}({OptionsInstanceVariableName}, {JsonMetadataServicesClassName}.{typeFriendlyName}Converter);";
+
+            return GenerateForType(typeMetadata, metadataInitSource);
+        }
+
+        private string GenerateForTypeWithUnknownConverter(TypeMetadata typeMetadata)
+        {
+            string typeCompilableName = typeMetadata.CompilableName;
+            string typeFriendlyName = typeMetadata.FriendlyName;
+
+            StringBuilder sb = new();
+
+            string metadataInitSource = $@"JsonConverter converter = {typeMetadata.ConverterInstantiationLogic};
+                    // TODO: consider moving this verification source to common helper.
```

Comment:
nit: spacing

Reply:
The spacing is correct given where the source should go in the generated file.